ipamd/ipamd.go

```diff
@@ -348,177 +352,254 @@ func (c *IPAMContext) StartNodeIPPoolManager() {
-	ipamdActionsInprogress.WithLabelValues("retryAllocENIIP").Add(float64(1))
-	defer ipamdActionsInprogress.WithLabelValues("retryAllocENIIP").Sub(float64(1))
+func (c *IPAMContext) decreaseIPPool(interval time.Duration) {
+	ipamdActionsInprogress.WithLabelValues("decreaseIPPool").Add(float64(1))
+	defer ipamdActionsInprogress.WithLabelValues("decreaseIPPool").Sub(float64(1))
 
-	curIPTarget, warmIPTargetDefined := c.getCurWarmIPTarget()
-	if warmIPTargetDefined && curIPTarget <= 0 {
-		log.Debugf("Skipping retry allocating ENI IP, warm IP target reached")
+	now := time.Now()
+	timeSinceLast := now.Sub(c.lastDecreaseIPPool)
+	if timeSinceLast <= interval {
+		log.Debugf("decreaseIPPool: skipping because time since last %v <= %v", timeSinceLast, interval)
 		return
 	}
-	maxIPLimit, err := c.awsClient.GetENIipLimit()
-	if err != nil {
-		log.Infof("Failed to retrieve ENI IP limit: %v", err)
-		return
-	}
-	eni := c.dataStore.GetENINeedsIP(maxIPLimit, UseCustomNetworkCfg())
-	if eni != nil {
-		log.Debugf("Attempt again to allocate IP address for ENI :%s", eni.ID)
-		var err error
-		if warmIPTargetDefined {
-			err = c.awsClient.AllocIPAddresses(eni.ID, int64(curIPTarget))
-		} else {
-			err = c.awsClient.AllocIPAddresses(eni.ID, maxIPLimit)
-		}
-		if err != nil {
-			ipamdErrInc("retryAllocENIIPAllocAllIPAddressFailed", err)
-			log.Warn("During eni repair: error encountered on allocate IP address", err)
-			return
-		}
-		ec2Addrs, _, err := c.getENIaddresses(eni.ID)
+
+	log.Debugf("Start decreasing IP pool")
+
+	// TODO(nckturner) shouldn't we attempt to remove more than 1 ENI at a time here?
+	eni := c.dataStore.RemoveUnusedENIFromDataStore()
+	if eni != "" {
+		log.Debugf("Start freeing ENI %s", eni)
+		err := c.awsClient.FreeENI(eni)
 		if err != nil {
-			ipamdErrInc("retryAllocENIIPgetENIaddressesFailed", err)
-			log.Warn("During eni repair: failed to get ENI ip addresses", err)
+			ipamdErrInc("decreaseIPPoolFreeENIFailed", err)
+			log.Errorf("Failed to free ENI %s, err: %v", eni, err)
 			return
 		}
-		c.lastNodeIPPoolAction = time.Now()
-		c.addENIaddressesToDataStore(ec2Addrs, eni.ID)
+	}
+	log.Info("No ENI to remove, all ENIs have IPs in use")
+
+	// Determine if there are IPs to free
+	// When we have extra IPs beyond the target and warmIPTargetDefined is enabled, deallocate extra IP addresses
+	if _, over, warmIPTargetDefined := c.ipTargetState(); warmIPTargetDefined && over > 0 {
+		eniInfos := c.dataStore.GetENIInfos()
+		for eniID := range eniInfos.ENIIPPools {
+			ips, err := c.findFreeableIPs(eniID)
+			if err != nil {
+				log.Errorf("Error finding unassigned IPs: %s", err)
+				return
+			}
+
+			if len(ips) == 0 {
+				continue
+			}
+
+			// Deallocate IPs from the instance if they aren't used by pods.
+			if err := c.awsClient.DeallocIPAddresses(eniID, ips); err != nil {
+				log.Debugf(fmt.Sprintf("Failed to decrease IP pool by removing IPs %v from ENI %s: %s", ips, eniID, err))
```

Comment:
What happens if we try to free 5 IPs and the call fails. Could 3 have been freed? Might be handled by the reconcile loop.